src/wp-includes/query.php

```diff
@@ -1148,17 +1161,42 @@ function _find_post_by_old_date( $post_type ) {
 		$date_query .= $wpdb->prepare( ' AND DAYOFMONTH(pm_date.meta_value) = %d', get_query_var( 'day' ) );
 	}
 
-	$id = 0;
 	if ( $date_query ) {
-		$id = (int) $wpdb->get_var( $wpdb->prepare( "SELECT post_id FROM $wpdb->postmeta AS pm_date, $wpdb->posts WHERE ID = post_id AND post_type = %s AND meta_key = '_wp_old_date' AND post_name = %s" . $date_query, $post_type, get_query_var( 'name' ) ) );
+		$key          = md5( $date_query );
+		$cache_key    = "_find_post_by_old_date:$key";
+		$cache        = wp_cache_get( $cache_key, 'posts' );
+
+		if ( $cache !== false ) {
+			return $cache;
+		} else {
+			$id = (int) $wpdb->get_var( $wpdb->prepare( "SELECT post_id FROM $wpdb->postmeta AS pm_date, $wpdb->posts WHERE ID = post_id AND post_type = %s AND meta_key = '_wp_old_date' AND post_name = %s" . $date_query, $post_type, get_query_var( 'name' ) ) );
+
+			if ( ! $id ) {
+				// Check to see if an old slug matches the old date.
+				$id = (int) $wpdb->get_var( $wpdb->prepare( "SELECT ID FROM $wpdb->posts, $wpdb->postmeta AS pm_slug, $wpdb->postmeta AS pm_date WHERE ID = pm_slug.post_id AND ID = pm_date.post_id AND post_type = %s AND pm_slug.meta_key = '_wp_old_slug' AND pm_slug.meta_value = %s AND pm_date.meta_key = '_wp_old_date'" . $date_query, $post_type, get_query_var( 'name' ) ) );
+			}
+			/**
+			 * Filters the old slug redirect post ID.
+			 *
+			 * @since 4.9.3
+			 *
+			 * @param int $id The redirect post ID.
+			 */
+			$id = apply_filters( 'old_slug_redirect_post_id', $id );
+
+			if ( ! $id ) {
+				return;
+			}
+
+			$link = get_permalink( $id );
+
+			wp_cache_set( $cache_key, $link, 'posts' );
```

Comment:
```suggestion
			wp_cache_set( $cache_key, $id, 'posts' );
```

Again, we should only be caching the ids here not the link.